// @polar-sh/sdk/dist/esm/models/components/benefitgrantlicensekeysproperties.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v4-mini";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const BenefitGrantLicenseKeysProperties$inboundSchema = z.pipe(z.object({
    license_key_id: z.optional(z.string()),
    display_key: z.optional(z.string()),
}), z.transform((v) => {
    return remap$(v, {
        "license_key_id": "licenseKeyId",
        "display_key": "displayKey",
    });
}));
/** @internal */
export const BenefitGrantLicenseKeysProperties$outboundSchema = z.pipe(z.object({
    licenseKeyId: z.optional(z.string()),
    displayKey: z.optional(z.string()),
}), z.transform((v) => {
    return remap$(v, {
        licenseKeyId: "license_key_id",
        displayKey: "display_key",
    });
}));
export function benefitGrantLicenseKeysPropertiesToJSON(benefitGrantLicenseKeysProperties) {
    return JSON.stringify(BenefitGrantLicenseKeysProperties$outboundSchema.parse(benefitGrantLicenseKeysProperties));
}
export function benefitGrantLicenseKeysPropertiesFromJSON(jsonString) {
    return safeParse(jsonString, (x) => BenefitGrantLicenseKeysProperties$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BenefitGrantLicenseKeysProperties' from JSON`);
}
