// @polar-sh/sdk/dist/esm/models/components/validationerror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v4-mini";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
/** @internal */
export const Loc$inboundSchema = smartUnion([
    z.string(),
    z.int(),
]);
/** @internal */
export const Loc$outboundSchema = smartUnion([
    z.string(),
    z.int(),
]);
export function locToJSON(loc) {
    return JSON.stringify(Loc$outboundSchema.parse(loc));
}
export function locFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Loc$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Loc' from JSON`);
}
/** @internal */
export const ValidationError$inboundSchema = z.object({
    loc: z.array(smartUnion([z.string(), z.int()])),
    msg: z.string(),
    type: z.string(),
});
/** @internal */
export const ValidationError$outboundSchema = z.object({
    loc: z.array(smartUnion([z.string(), z.int()])),
    msg: z.string(),
    type: z.string(),
});
export function validationErrorToJSON(validationError) {
    return JSON.stringify(ValidationError$outboundSchema.parse(validationError));
}
export function validationErrorFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ValidationError$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ValidationError' from JSON`);
}
